### PR TITLE
Refactor sample application

### PR DESCRIPTION
**Pull request by olbjo**
We should rerun Auomated RT-tester suite on both rt-kernel and linux targets before merging. 

Commit message:
Split sample application into several files.
- Move utilities to separate file
- Add log macros for sample application
- Move gsdml parameters to specific header
- Move process data and parameters to separate file
- Add unique submodule ids
- GSDML has changed. Needs to be reloaded in engineering tool and
  ART tester.
- Align sampleapp_main for linux and rt-kernel

### CODE REVIEW

File: sample_app/app_gsdml.h [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+#define APP_GSDML_DIAG_CUSTOM_USI 0x1234
+
+/* See "Specification for GSDML" 8.26 LogBookEntryItem for allowed values */
+#define APP_GSDML_LOGBOOK_ERROR_CODE   0x20 /* Manufacturer specific */
+#define APP_GSDML_LOGBOOK_ERROR_DECODE 0x82 /* Manufacturer specific */
+#define APP_GSDML_LOGBOOK_ERROR_CODE_1 PNET_ERROR_CODE_1_FSPM
+#define APP_GSDML_LOGBOOK_ERROR_CODE_2 0x00       /* Manufacturer specific */
+#define APP_GSDML_LOGBOOK_ENTRY_DETAIL 0xFEE1DEAD /* Manufacturer specific */
+
+#define APP_GSDM_GSDML_IDX_1  123
```
Comment (pyhys) [inline sample_app/app_gsdml.h:76]: `APP_GSDML_IDX_1` ?

File: sample_app/GSDML-V2.4-RT-Labs-P-Net-Sample-App-20210415.xml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="iso-8859-1"?>
```
Comment (pyhys) [inline sample_app/GSDML-V2.4-RT-Labs-P-Net-Sample-App-20210415.xml:1]: Rename GSDML file with new date

File: sample_app/app_data.h [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+   uint32_t submodule_id,
+   uint32_t index,
+   uint8_t ** data,
+   uint16_t * length);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* APP_DATA_H */
```
Comment (pyhys) [inline sample_app/app_data.h:123]: Missing newline

File: sample_app/app_gsdml.c [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            if (app_gsdml_parameters[j].index == index)
+            {
+               return &app_gsdml_parameters[j];
+            }
+         }
+      }
+   }
+
+   return NULL;
+}
```
Comment (pyhys) [inline sample_app/app_gsdml.c:86]: missing newline

File: sample_app/app_gsdml.h [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+/**
+ * @file
+ * @brief Device properties defined by the gsdml device definition
+ *
+ * Functions for getting module, submodule and parameter
+ * configurations using their ids.
+ *
+ * Important:
+ * Any change in this file may require an update of the gsdml file.
```
Comment (pyhys) [inline sample_app/app_gsdml.h:27]: GSDML in upper case

File: sample_app/app_gsdml.h [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+ */
+const app_gsdml_param_t * app_gsdml_get_parameter_cfg (
+   uint32_t submodule_id,
+   uint32_t index);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* APP_GSDML_H */
```
Comment (pyhys) [inline sample_app/app_gsdml.h:284]: Missing newline

File: sample_app/app_log.c [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ * full license information.
+ ********************************************************************/
+
+#include "app_log.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdarg.h>
```
Comment (pyhys) [inline sample_app/app_log.c:22]: Sort alphabetically if possible

File: sample_app/app_utils.c [hunk 8]
```diff
@@ -0,0 +1,714 @@
...
+      break;
+   case PNET_IOXS_GOOD:
+      s = "IOXS_GOOD";
+      break;
+   }
+
+   return s;
+}
+
+void app_utils_get_error_code_strings (
```
Comment (pyhys) [inline sample_app/app_utils.c:100]: Great!

File: sample_app/app_utils.c [hunk 9]
```diff
@@ -0,0 +1,714 @@
...
+   cfg->oem_device_id.device_id_hi = 0xee;
+   cfg->oem_device_id.device_id_lo = 0x01;
+   strcpy (cfg->product_name, APP_GSDML_PRODUCT_NAME);
+
+   /* Timing */
+   cfg->min_device_interval = APP_GSDML_MIN_DEVICE_INTERVAL;
+
+   /* Network configuration */
+   cfg->send_hello = true;
+   cfg->num_physical_ports = 1;
```
Comment (pyhys) [inline sample_app/app_utils.c:263]: Is this later overwritten? If so, add a comment.

File: sample_app/app_utils.c [hunk 10]
```diff
@@ -0,0 +1,714 @@
...
+
+bool app_utils_subslot_is_input (const app_subslot_t * p_subslot)
+{
+   if (
+      p_subslot != NULL && (p_subslot->data_cfg.data_dir == PNET_DIR_INPUT ||
+                            p_subslot->data_cfg.data_dir == PNET_DIR_IO))
+   {
+      return true;
+   }
+   else
```
Comment (pyhys) [inline sample_app/app_utils.c:650]: Maybe replace 
```
else
{
   return false;
}
```
with
```

return false;
```
(note blank line)
Also for the other functions

File: sample_app/app_gsdml.h [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+ * Design requires unique submodule ids.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <pnet_api.h>
+
+#define APP_GSDM_API 0
```
Comment (pyhys) [inline sample_app/app_gsdml.h:40]: `APP_GSDML_API`

File: sample_app/sampleapp_common.h [hunk 12]
```diff
@@ -24,165 +24,18 @@
...
-#define CHANNEL_ERRORTYPE_LINE_BREAK                          0x0006
-#define CHANNEL_ERRORTYPE_DATA_TRANSMISSION_IMPOSSIBLE        0x8000
-#define CHANNEL_ERRORTYPE_NETWORK_COMPONENT_FUNCTION_MISMATCH 0x8008
-#define EXTENDED_CHANNEL_ERRORTYPE_FRAME_DROPPED              0x8000
-#define EXTENDED_CHANNEL_ERRORTYPE_MAUTYPE_MISMATCH           0x8001
-#define EXTENDED_CHANNEL_ERRORTYPE_LINE_DELAY_MISMATCH        0x8002
+#define APP_TICK_INTERVAL_US 1000 /* 1 ms */
 
-/********************** Settings **********************************************/
+#define APP_MAIN_THREAD_PRIORITY  15
```
Comment (pyhys) [inline sample_app/sampleapp_common.h:29]: Are these two for Linux?

**Review by pyhys — COMMENTED** [review]
Great!
I will look in more detail.

File: sample_app/app_gsdml.h (hunk 1 above)
Comment (olbjo) [reply sample_app/app_gsdml.h:76]: OK
APP_GSDML_PARAMETER_1_IDX
APP_GSDML_PARAMETER_2_IDX
APP_GSDML_PARAMETER_LENGTH

File: sample_app/GSDML-V2.4-RT-Labs-P-Net-Sample-App-20210415.xml (hunk 2 above)
Comment (olbjo) [reply sample_app/GSDML-V2.4-RT-Labs-P-Net-Sample-App-20210415.xml:1]: OK

File: sample_app/app_data.h (hunk 3 above)
Comment (olbjo) [reply sample_app/app_data.h:123]: OK

File: sample_app/app_gsdml.c (hunk 4 above)
Comment (olbjo) [reply sample_app/app_gsdml.c:86]: OK

File: sample_app/app_gsdml.h (hunk 5 above)
Comment (olbjo) [reply sample_app/app_gsdml.h:27]: OK

File: sample_app/app_gsdml.h (hunk 6 above)
Comment (olbjo) [reply sample_app/app_gsdml.h:284]: OK

File: sample_app/app_log.c (hunk 7 above)
Comment (olbjo) [reply sample_app/app_log.c:22]: OK, most of them should not be there anyway

File: sample_app/app_utils.c (hunk 8 above)
Comment (olbjo) [reply sample_app/app_utils.c:100]: We should add more error codes but thats for another commit

File: sample_app/app_utils.c (hunk 9 above)
Comment (olbjo) [reply sample_app/app_utils.c:263]: OK
Updated docstring in header and added comment of this line

File: sample_app/app_utils.c (hunk 10 above)
Comment (olbjo) [reply sample_app/app_utils.c:650]: OK

File: sample_app/app_gsdml.h (hunk 11 above)
Comment (olbjo) [reply sample_app/app_gsdml.h:40]: OK

File: sample_app/sampleapp_common.h (hunk 12 above)
Comment (olbjo) [reply sample_app/sampleapp_common.h:29]: Yes, added comment

File: sample_app/sampleapp_common.c [hunk 13]
```diff
@@ -1858,80 +1210,85 @@ static void app_handle_send_alarm (
...
             p_subslot->subslot_nbr,
             APP_ALARM_USI,
-            APP_ALARM_PAYLOAD_SIZE,
-            alarm_payload);
-         *p_alarm_allowed = false; /* Not allowed until ACK received */
+            APP_GSDM_ALARM_PAYLOAD_SIZE,
+            app->alarm_payload);
+         app->alarm_allowed = false; /* Not allowed until ACK received */
+
+         // todo handle return code on pnet_alarm_send_process_alarm
```
Comment (pyhys) [inline sample_app/sampleapp_common.c:1244]: Use C style comment `/* */`

File: sample_app/sampleapp_common.h [hunk 14]
```diff
@@ -195,224 +52,87 @@ struct cmd_args
...
-   uint16_t slot_nbr;
-   uint16_t subslot_nbr;
-   uint32_t submodule_id;
-   const char * submodule_name;
-   pnet_data_cfg_t data_cfg;
-   uint8_t * p_in_data;
-} app_subslot_t;
+   RUN_IN_SEPARATE_THREAD,
+   RUN_IN_MAIN_THREAD
+} app_run_in_separate_task;
```
Comment (pyhys) [inline sample_app/sampleapp_common.h:61]: `app_run_in_separate_task_t` ?

File: sample_app/GSDML-V2.4-RT-Labs-P-Net-Sample-App-20210507.xml [hunk 15]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="iso-8859-1"?>
-<ISO15745Profile xmlns="http://www.profibus.com/GSDML/2003/11/DeviceProfile" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.profibus.com/GSDML/2003/11/DeviceProfile ..\xsd\GSDML-DeviceProfile-V2.4.xsd">
+<ISO15745Profile xmlns="http://www.profibus.com/GSDML/2003/11/DeviceProfile"
+   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.profibus.com/GSDML/2003/11/DeviceProfile ..\xsd\GSDML-DeviceProfile-V2.4.xsd">
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/GSDML-V2.4-RT-Labs-P-Net-Sample-App-20210507.xml:3]: OK

File: sample_app/app_data.c [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+      }
+   }
+   return -1;
+}
+
+int app_data_set_default_outputs (void)
+{
+   if (outputdata[0] != APP_DATA_DEFAULT_OUTPUT_DATA)
+   {
+      APP_LOG_DEBUG ("Setting outputs to default values.\n");
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/app_data.c:130]: OK

File: sample_app/app_gsdml.h [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+ * @file
+ * @brief Device properties defined by the GSDML device definition
+ *
+ * Functions for getting module, submodule and parameter
+ * configurations using their ids.
+ *
+ * Important:
+ * Any change in this file may require an update of the GSDML file.
+ * Note that when the GSDML file is updated it has to be reloaded
+ * in your Profinet engineering to and PLC applications may be affected.
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/app_gsdml.h:29]: OK

File: sample_app/app_gsdml.h [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+
+#define APP_GSDML_INPUT_DATA_SIZE   1 /* bytes, for digital inputs data */
+#define APP_GSDML_OUTPUT_DATA_SIZE  1 /* bytes, for digital outputs data */
+#define APP_GSDM_ALARM_PAYLOAD_SIZE 1 /* bytes */
+
+static const app_gsdml_module_t dap_1 = {
+   .id = PNET_MOD_DAP_IDENT,
+   .name = "DAP 1",
+   .submodules =
+      {PNET_SUBMOD_DAP_IDENT, PNET_SUBMOD_DAP_INTERFACE_1_PORT_1_IDENT, 0}};
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/app_gsdml.h:120]: OK

File: sample_app/app_log.h [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+#include <stdint.h>
+
+#define APP_LOG_LEVEL_DEBUG   0x00
+#define APP_LOG_LEVEL_INFO    0x01
+#define APP_LOG_LEVEL_WARNING 0x02
+#define APP_LOG_LEVEL_ERROR   0x03
+#define APP_LOG_LEVEL_FATAL   0x04
+
+#define PNU_DEFAULT_LOG_LEVEL APP_LOG_LEVEL_FATAL
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/app_log.h:39]: OK! APP_

File: sample_app/app_utils.h [hunk 20]
```diff
@@ -0,0 +1,373 @@
...
+ * Convert error code to string format
+ * Only common error codes supported.
+ * Todo: Add rest of error codes.
+ *
+ * @param err_cls        In:   The error class. See PNET_ERROR_CODE_1_*
+ * @param err_code       In:   The error code. See PNET_ERROR_CODE_2_*
+ * @param err_cls_str    Out:   The error class string
+ * @param err_code_str   Out:   The error code string
+ */
+
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/app_utils.h:144]: OK

File: sample_app/app_utils.h [hunk 21]
```diff
@@ -0,0 +1,373 @@
...
+ * @param event            In:    event
+ * @return  A string representing the event
+ */
+const char * app_utils_event_to_string (pnet_event_values_t event);
+
+/**
+ * Get network configuration from a string
+ * defining a list of network interfaces examples:
+ * "eth0" or "br0,eth0,eth1"
+ * @param netif_list_str      In: list of network ifs
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/app_utils.h:180]: OK

File: sample_app/app_utils.h [hunk 22]
```diff
@@ -0,0 +1,373 @@
...
+ * Examples:
+ * arg_str                 num_ports
+ * "eth0"                  1
+ * "eth0,eth1"             error (We need a bridge as well)
+ * "br0,eth0,eth1"         2
+ *
+ * @param arg_str      In:   Network interface list as comma separated,
+ *                           terminated string. For example "eth0" or
+ *                           "br0,eth0,eth1".
+ * @param max_port     In:   PNET_MAX_PHYSICAL_PORTS, passed as argument to
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/app_utils.h:213]: This has not changed I should be ok.

File: sample_app/sampleapp_common.c [hunk 23]
```diff
@@ -1324,471 +863,296 @@ void app_plug_dap (pnet_t * net, void * arg)
...
+ * @param tag        In:   Application handle
  */
-static void app_handle_cyclic_data (
-   pnet_t * net,
-   const app_data_t * p_appdata,
-   bool button_pressed,
-   uint8_t data_ctr,
-   uint8_t * p_inputdata,
-   uint16_t inputdata_size)
+
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/sampleapp_common.c:929]: OK

File: sample_app/sampleapp_common.c [hunk 24]
```diff
@@ -1809,29 +1173,17 @@ static void app_handle_cyclic_data (
...
-   pnet_t * net,
-   uint32_t arep,
-   bool * p_alarm_allowed,
-   const app_data_t * p_appdata,
-   uint8_t * alarm_payload)
+static void app_handle_demo_pnet_api (app_data_t * app)
 {
-   static app_demo_state_t state = APP_DEMO_STATE_ALARM_SEND;
    uint16_t slot = 0;
    bool found_inputslot = false;
```
[the remark this answers is not in the document]
Comment (olbjo) [reply sample_app/sampleapp_common.c:1181]: OK

File: sample_app/sampleapp_common.c (hunk 13 above)
Comment (olbjo) [reply sample_app/sampleapp_common.c:1244]: OK

File: sample_app/sampleapp_common.h (hunk 14 above)
Comment (olbjo) [reply sample_app/sampleapp_common.h:61]: OK

**Review by hefloryd — COMMENTED** [review]
How much of the "app"-code is generic enough that it could be part of the p-net stack?

**Review by hefloryd — APPROVED** [review]

**Comment by olbjo** [comment]
> How much of the "app"-code is generic enough that it could be part of the p-net stack?

I think the app_utils files except for the initialization of the pnet config could be moved to the stack. 
The gsdml part can be done in different ways. I chose something that is easy to follow. Depending on the complexity of the  device a user may want to add another abstraction layer here. I do not think gsdml part should be moved into the stack.

**Pull request merged**